staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go

```diff
@@ -0,0 +1,271 @@
+		err = fmt.Errorf("contentType=%s, expected application/json", contentType)
+		klog.ErrorS(err, "unable to process a request with an unknown content type", "content type", contentType)
+		http.Error(w, "unable to process a request with a non-json content type", http.StatusBadRequest)
+		return
+	}
+
+	v1AdmissionReviewKind := admissionv1.SchemeGroupVersion.WithKind("AdmissionReview")
+	reviewObject, gvk, err := codecs.UniversalDeserializer().Decode(body, &v1AdmissionReviewKind, nil)
+	if err != nil {
+		klog.ErrorS(err, "unable to decode the request")
+		http.Error(w, "unable to decode the request", http.StatusBadRequest)
+		return
+	}
+	if *gvk != v1AdmissionReviewKind {
+		klog.InfoS("Unexpected AdmissionReview kind", "kind", gvk.String())
+		http.Error(w, fmt.Sprintf("unexpected AdmissionReview kind: %s", gvk.String()), http.StatusBadRequest)
+		return
+	}
+	review, ok := reviewObject.(*admissionv1.AdmissionReview)
+	if !ok {
+		klog.InfoS("Failed admissionv1.AdmissionReview type assertion")
+		http.Error(w, "unexpected AdmissionReview type", http.StatusBadRequest)
+	}
+	klog.V(1).InfoS("received request", "UID", review.Request.UID, "kind", review.Request.Kind, "resource", review.Request.Resource)
+
+	attributes := admission.RequestAttributes(review.Request, codecs.UniversalDeserializer())
+	response := s.delegate.Validate(ctx, attributes)
+	response.UID = review.Request.UID // Response UID must match request UID
+	review.Response = response
+	writeResponse(w, review)
+}
+
+// Config holds the loaded options.Options used to set up the webhook server.
+type Config struct {
+	SecureServing     *apiserver.SecureServingInfo
+	InsecureServing   *apiserver.DeprecatedInsecureServingInfo
+	KubeConfig        *restclient.Config
+	PodSecurityConfig *admissionapi.PodSecurityConfiguration
+}
+
+// LoadConfig loads the Config from the Options.
+func LoadConfig(opts *options.Options) (*Config, error) {
+	if errs := opts.Validate(); len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+
+	var c Config
+	opts.SecureServing.ApplyTo(&c.SecureServing)
+	opts.InsecureServing.ApplyTo(&c.InsecureServing)
+
+	// Load Kube Client
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", opts.KubeConfigFile)
+	if err != nil {
+		return nil, err
+	}
+	c.KubeConfig = restclient.AddUserAgent(kubeConfig, "podsecurity-webhook")
+
+	// Load PodSecurity config
+	c.PodSecurityConfig, err = podsecurityconfigloader.LoadFromFile(opts.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return &c, nil
+}
+
+// Setup creates an Admission object to handle the admission logic.
+func Setup(c *Config) (*Server, error) {
+	s := &Server{
+		secureServing:   c.SecureServing,
+		insecureServing: c.InsecureServing,
+	}
+
+	if s.secureServing == nil && s.insecureServing == nil {
+		return nil, errors.New("no serving info configured")
+	}
+
+	client, err := clientset.NewForConfig(c.KubeConfig)
```

Comment:
what's the QPS/Burst for this client? because this is doing live lookups for pods when updating a namespace, and falling back to live lookups for namespaces not in the cache, client-side throttling here can delay admission requests

because there's no authn/authz before processing an admission request, a pod admission request for a bogus namespace could DOS the client